sec/security.go

```diff
@@ -24,3 +24,17 @@ type SecureTransport interface {
 	// SecureOutbound secures an outbound connection.
 	SecureOutbound(ctx context.Context, insecure net.Conn, p peer.ID) (SecureConn, error)
 }
+
+// A SecureMuxer is a wrapper around SecureTransport which can select security protocols
+// and open outbound connections with simultaneous open.
+type SecureMuxer interface {
+	// SecureInbound secures an inbound connection.
+	// The returned boolean indicates whether the connection should be trated as a server
+	// connection; in the case of SecureInbound it should always be true.
+	SecureInbound(ctx context.Context, insecure net.Conn) (SecureConn, bool, error)
+
+	// SecureOutbound secures an outbound connection.
+	// The returned boolean indicates whether the connection should be treated as a server
+	// connection due to simultaneous open.
+	SecureOutbound(ctx context.Context, insecure net.Conn, p peer.ID) (SecureConn, bool, error)
+}
```

Comment:
I'd like to consider using a special sentinel error instead. That is, return a connection _and_ a special `ErrSimultaniousConnect`.

This way we can avoid polluting the interfaces too much.

Reply:
What would the error buy is? We can still secure the connection with simultaneous open.

Reply:
also, simultaneous connect is _not_ an error!

Reply:
I'm just trying to avoid having this TCP specific feature from leaking into the interface.

Reply:
We discussed this synchronously. I was assuming we'd be able to come up with a better interface the next time we add a transport from scratch. However, @vyzo convinced me otherwise.

The only viable alternative is to add a third `Secure(ctx, insecure, p) (Conn, iAmInitiator bool, error)` function. However, that has no real benefit over `SecureOutbound`

Well, I guess it has a _slight_ benefit: Transports that _aren't_ TCP could skip this whole dance. Maybe _that's_ worth it @vyzo? Regardless, my hope to avoid messing with the interfaces seems moot.